### PR TITLE
.NET GitHub action

### PR DESCRIPTION
**Pull request by IEvangelist**
## Summary

A Docker containerized .NET app, that serves as a GitHub action. When ran will analyze C# and VB projects for code metrics, reporting:

- Cyclomatic complexity
- Maintainability index
- Class coupling
- Depth of inheritance
- Exact number of lines of code
- Approximated number of lines of executable code

It writes a *CODE_METRICS.md* file at the root directory given, and will enable GitHub action workflow composition to create PRs. In other words, consuming this GitHub action with the create-pull-request action in a .NET repo will provide automatic PRs to maintain the code metrics as a markdown file.

Ignore the `DotNet.CodeAnalysis` project in the review it is mostly borrowed source, instead focus on the `DotNet.GitHubAction` bits.

Contributes to dotnet/docs#23117

### CODE REVIEW

File: github-actions/DotNet.GitHubAction/DotNet.CodeAnalysis/FileNameUtilities.cs [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+        }
+
+        /// <summary>
+        /// Returns the offset in <paramref name="path"/> where the dot that starts an extension is, or -1 if the path doesn't have an extension.
+        /// </summary>
+        /// <remarks>
+        /// Returns 0 for path ".goo".
+        /// Returns -1 for path "goo.".
+        /// </remarks>
+        private static int IndexOfExtension(string? path)
```
Comment (BillWagner) [inline github-actions/DotNet.GitHubAction/DotNet.CodeAnalysis/FileNameUtilities.cs:40]: I think much of the logic in this method could be simplified using `LastIndexOf`.

Comment (BillWagner) [reply github-actions/DotNet.GitHubAction/DotNet.CodeAnalysis/FileNameUtilities.cs:40]: In a comment, you should make it clear what the result for this method would be in this cases:

- ".bashrc" (any file that starts with "."
- "application.config.json" (any file that contains multiple "." characters.

File: github-actions/DotNet.GitHubAction/DotNet.CodeAnalysis/FileUtilities.cs [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+                if (fileExists(combinedPath))
+                {
+                    return combinedPath;
+                }
+            }
+
+            return null;
+        }
+
+        internal static string? ResolveRelativePath(string? path, string? baseDirectory)
```
Comment (BillWagner) [inline github-actions/DotNet.GitHubAction/DotNet.CodeAnalysis/FileUtilities.cs:84]: I'm a fan of expression bodied members here, but that's a style issue and a nit.

File: github-actions/DotNet.GitHubAction/README.md [hunk 3]
```diff
@@ -0,0 +1,8 @@
+<!--
```
Comment (BillWagner) [inline github-actions/DotNet.GitHubAction/README.md:1]: This doesn't look completed.

I'm fine merging for now, but I think this gets onboarded once we merge.

**Review by BillWagner — APPROVED** [review]
This really looks great @IEvangelist 

I had a few optional comments, and then you can :shipit:

My main concern is that this sample will get onboarded before you update the README file.

File: github-actions/DotNet.GitHubAction/DotNet.CodeAnalysis/FileUtilities.cs [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+        {
+            return ResolveRelativePath(path, null, baseDirectory);
+        }
```
Comment (IEvangelist) [inline github-actions/DotNet.GitHubAction/DotNet.CodeAnalysis/FileUtilities.cs:87]:
```suggestion
        internal static string? ResolveRelativePath(string? path, string? baseDirectory)
            => ResolveRelativePath(path, null, baseDirectory);
```

**Pull request merged**